keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,651 @@
+
+### Non-Goals
+
+Removing cgroup v1 support immediately. The deprecation and removal process will be gradual, allowing users ample time to migrate.
+
+## Proposal
+
+The proposal outlines a plan to deprecate and then remove cgroup v1 support in Kubernetes, encouraging the community and users to transition to cgroup v2.
+
+### Risks and Mitigations
+
+The main risk involves potential disruptions to users who have not migrated to cgroup v2. Mitigations include providing clear documentation on the migration process, offering community support for common issues encountered during migration, and maintaining cgroup v1 support for a predefined deprecation period to allow for a smoother transition.
+
+Users depending following technologies will have to use their respective listed versions below:
+
+- OpenJDK / HotSpot: jdk8u372, 11.0.16, 15 and later
+
+- NodeJs 20.3.0 or later
+
+- IBM Semeru Runtimes: jdk8u345-b01, 11.0.16.0, 17.0.4.0, 18.0.2.0 and later
+
+- IBM SDK Java Technology Edition Version (IBM Java): 8.0.7.15 and later
+
+- If users run any third-party monitoring and security agents that depend on the cgroup file system, they need to update the agents to a version that supports cgroup v2.
+
+## Design Details
+
+## Handling Deprecation Warnings and Migration Path Flexibility
+
+### Introduction of cgroup version metric
+
+A new metric, `kubelet_cgroup_version`, is proposed. This metric will report values `1` or `2`, indicating whether the host is utilizing cgroup version `1` or `2`, respectively. The kubelet will assess the host's cgroup version at startup and emit this metric accordingly.
+
+The introduction of this metric aims to streamline the process for cluster administrators in determining the cgroup version deployed across their hosts. This metric removes the need for manual node inspection, providing a clear insight into the cgroup version each node operates on.
+
+### Implementing a warning log and an event for cgroup v1 usage
+
+Starting from 1.31, during kubelet startup if the host is running on cgroup v1, kubelet will log a warning message like,
+
+```golang
+klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+```
+and also emit a corresponding event,
+```golang
+eventRecorder.Event(pod, v1.EventTypeWarning, "RemovedCgroupV1", fmt.Sprint("cgroup v1 support has been deprecated"))
```

Comment:
Why `RemovedCgroupV1`?

Reply:
https://github.com/kubernetes/enhancements/pull/4572#discussion_r1554395982

Reply:
Whatever we call the feature gate, I would use a different reason code. For example, one of:
- `Deprecation`
- `CGroupSupportDeprecation`
- `DeprecatedConfigurationDetected`

Other names are fine too; the key detail is that the reason should explain why the node is reporting an Event, not some future change to Kubernetes. _Future_ because clearly the v1 cgroup support won't have been removed at the point this event is logged.

Also, think about the event that we'll fire once the support is actually removed; for example: `UnsupportedConfiguration`, `CGroupVersionIncompatibility`, or the lighthearted `ReadTheUpgradeNotes` (don't actually use that last one).

Reply:
>ReadTheUpgradeNotes (don't actually use that last one). 

:)